### DISCUSSION TITLE
Using fromEvent with SignalR hub connection.

### DISCUSSION THREAD

**Discussion by kosmakoff** [ReactiveX/rxjs#6152] [Q&A] [answered]
I am integrating RxJS with SignalR. SignalR has this type called `HubConnection` which in turn has 2 jQuery-like events subscription methods `on` and `off`.

I tried to pass the instance of `HubConnection` directly to `fromEvent` function to create an observable of this connection's events.

The issue is that Typescript won't allow that, because `fromEvent` expects the handler's type is defined as `Function` in rxjs, as follows:
```ts
export interface JQueryStyleEventEmitter {
    on: (eventName: string, handler: Function) => void;
    off: (eventName: string, handler: Function) => void;
}
```

Whereas the connections' `on` method signature looks like this:
```ts
on(methodName: string, newMethod: (...args: any[]) => void): void;
```

Typescript complains about type `Function` doesn't fit into signature `(...args: any[]): void`. I'm not that proficient in Typescript to understand whether it is a limitation of Typescript, or shortcoming of RxJS itself.

Any advise, anyone, please?

**Comment by voliva** [comment, marked as answer]
Yep... it seems like there's something missmatching on the types of Function.

However, you have an easy workaround by using `fromEventPattern`:

```ts
const fromSignalREvent = (connection: HubConnection, eventName: string) =>
  fromEventPattern(
    handler => connection.on(eventName, handler),
    handler => connection.off(eventName, handler)
  )
```

**Reply by cartant** [reply]
FWIW, the v7 types won't be using `Function`:

https://github.com/ReactiveX/rxjs/blob/105bf99d27144a24100b211116af1341c5263738/src/internal/observable/fromEvent.ts#L30-L33

`Function` is [one of those types that's best avoided completely](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md#default-options).
